### PR TITLE
`samplerd` service

### PR DESCRIPTION
**Pull request by goodboy**
Adds a new singleton `samplerd` actor to our service tree which provides:

- single process sample incrementing for all registered `ShmArray` buffers thereby providing a unified and synchronized "step update" for all real-time flows.
- single process-task for all index-step broadcasting ensuring all consumers are seeing the same update msgs in tandem.

The overall service tree should now look something like:

```bash
pikerd /home/goodboy/.virtualenvs/piker310/bin/pikerd --tsdb --pdb
  ├─python3 -m tractor._child --uid ('marketstored', '60ab7cd9-819e-4fb8-95ce-ea343ef1aa81') --parent_addr('12
  │   └─3*[{python3}]
  ├─python3 -m tractor._child --uid ('brokerd.binance', '12de17a0-5e63-4083-879d-45772647dd68') --parent_addr(
  │   └─3*[{python3}]
  ├─python3 -m tractor._child --uid ('brokerd.ib', 'ec8fb1bc-8dd0-4317-8bd3-f615a6d7c255') --parent_addr('127.
  │   └─4*[{python3}]
  ├─python3 -m tractor._child --uid ('samplerd', '42107775-c146-4589-8aaa-7aa170faa618') --parent_addr('127.0.
  │   └─3*[{python3}]
  ├─python3 -m tractor._child --uid ('emsd', 'ecd5fdbc-aa05-404a-b738-31056b9c24dd') --parent_addr('127.0.0.1'
  │   ├─python3 -m tractor._child --uid ('paperboi.binance', '6b8833a2-6c20-494c-8de8-ff6b9d1015eb') --parent_addr
  │   │   └─3*[{python3}]
  │   └─3*[{python3}]
  └─3*[{pikerd}]
```
---
Further included enhancements in this patchset:
- simplified `._daemon.Services` API which presents as a singleton `class` with `@classmethod`s and is expected to exist once-and-only-once in the `pikerd` actor.
- improved quote feed throttling by framing ticks by type over a given throttle period resulting in lower transmission rates and a "update from last" msg format which greatly speeds up the graphics update loop.
  - now accomplished by a self-contained [framing helper func](https://github.com/pikers/piker/blob/samplerd_service/piker/data/_sampling.py#L695)
- misc fixes for history loading from the tsdb mostly to do with explicit query error logging and a guard to avoid inf looping on duplicate history frame detection. .
- syncing FSP flows with their sources by copying epoch time indexing from src to dst.
- misc broker backend fixes:
  - `kraken` fix to avoid ws unsubs on reconnect and adjustment to better parse `src_fiat` for position tracking.
  - `binance` always query for an extra 1Min ohlc bar to ensure **latest** sample step is being used.

---
What we might factor out:
- [x] fix to avoid cancelling the dark clearing loop on data feed `Lagged`s
  => **moved into #441** which requires https://github.com/goodboy/tractor/pull/343 from `tractor`
  
---
Further TODOs:
- [x] add some basic service tree tests
  - [x] verify a `samplerd` is always spawned
  - [x] verify an fsp cluster can be spawned and queried
  - [x] verify an `emsd` and `paperboi` can be allocated and queried

### CODE REVIEW

**Comment by goodboy** [comment]
Test suite in [fe4fb37](https://github.com/pikers/piker/pull/440/commits/fe4fb37b58bea28b8591746de079d7ba81878351) peepers!

This is ready to land after review 🏄🏼

**Comment by goodboy** [comment]
Bleh/lol that test is failing do to error msg differences in dev branch of tractor vs. this..

**Comment by goodboy** [comment]
There, now `tractor`'s branch `piker_pin` is in the right spot 🏄🏼

File: piker/data/feed.py [hunk 1]
```diff
@@ -405,14 +382,26 @@ async def start_backfill(
                 # request loop until the condition is resolved?
                 return
 
-            if next_start_dt in starts:
+            if (
+                next_start_dt in starts
+                and starts[next_start_dt] <= 6
```
Comment (algorandpa) [inline piker/data/feed.py:387]: magic number

File: piker/data/flows.py [hunk 2]
```diff
@@ -0,0 +1,321 @@
...
+
+    ) -> np.ndarray:
+        '''
+        Slice an input struct array providing only datums
+        "in view" of this chart.
+
+        '''
+        arr = {
+            1: self.rt_shm.array,
+            60: self.hist_shm.arry,
```
Comment (algorandpa) [inline piker/data/flows.py:253]: magic bois

**Review by algorandpa — APPROVED** [review]
A lot of stuff over my head but LGTM!

File: piker/clearing/_client.py [hunk 3]
```diff
@@ -213,14 +227,16 @@ async def open_ems(
...
 
-    mode: str = 'live'
-
     async with maybe_open_emsd(broker) as portal:
 
         mod = get_brokermod(broker)
-        if not getattr(mod, 'trades_dialogue', None):
+        if (
+            not getattr(mod, 'trades_dialogue', None)
+            or mode == 'paper'
```
[the remark this answers is not in the document]
Comment (goodboy) [reply piker/clearing/_client.py:235]: there's 2 cases:
- the backend doesn't support live trading (and the first predicate fails)
- the caller manually specifies paper mode (over some other option, presumably `'live'`)

File: piker/data/feed.py (hunk 1 above)
Comment (goodboy) [reply piker/data/feed.py:387]: yeah i mean, i just tend to use 6s and 1s and there's no reason. It's literally for random choices that the universe favors.

I mean why else would we try 6 times.. i dunno

File: piker/data/flows.py (hunk 2 above)
Comment (goodboy) [reply piker/data/flows.py:253]: seconds bruv, all sample periods are in seconds terms.

you think i should doc it better?

File: piker/data/feed.py [hunk 4]
```diff
@@ -309,6 +281,11 @@ async def start_backfill(
             - pendulum.from_timestamp(times[-2])
         ).seconds
 
+        if step_size_s == 60:
+            inow = round(time.time())
+            if (inow - times[-1]) > 60:
+                await tractor.breakpoint()
```
Comment (guilledk) [inline piker/data/feed.py:287]: Should we keep this `breakpoint`?

**Review by guilledk — APPROVED** [review]

File: piker/data/feed.py (hunk 4 above)
Comment (goodboy) [reply piker/data/feed.py:287]: good catch. no should probably not since i don't think this issue, when it shows up is a major problem.

**Pull request merged**